### ISSUE TITLE
Implement support for PowerPlans via the DiscernObjectFactory:POWERORDERS COM object

### ISSUE THREAD

**Issue by geekmdtravis** [geekmdtravis/fluent-cerner-js#69]
Task List
----

- [x] unit test functions and branches
- [x] update `demo.js`
- [x] update `README.MD`
- [x] ensure capture of valuable return data and transform it into more user-friendly return objects/properties
- [x] merge into branch `version-0.10` when complete

Power plans can be placed using this COM object and should be supported in the library.

**POWERORDERS**

- [AddDiagnosesToOrder](https://wiki.cerner.com/display/MPDEVWIKI/AddDiagnosesToOrder?src=contextnavchildmode) - medium utility. can be good to help keep orders clean. 
- [AddNewOrdersToScratchpad](https://wiki.cerner.com/display/MPDEVWIKI/AddNewOrdersToScratchpad?src=contextnavchildmode) - low utility
- [AddNewOrderToScratchpad](https://wiki.cerner.com/display/MPDEVWIKI/AddNewOrderToScratchpad?src=contextnavchildmode) - low utiliy
- [AddPowerPlanMOEW](https://wiki.cerner.com/display/MPDEVWIKI/AddPowerPlanMOEW?src=contextnavchildmode) - inferior to `AddPowerPlanWithDetails`
- [AddPowerPlanWithDetails](https://wiki.cerner.com/display/MPDEVWIKI/AddPowerPlanWithDetails?src=contextnavchildmode) - high utility
- [CreateMOEW](https://wiki.cerner.com/display/MPDEVWIKI/CreateMOEW?src=contextnavchildmode) - high utility as it supports other actions
- [DestroyMOEW](https://wiki.cerner.com/display/MPDEVWIKI/DestroyMOEW?src=contextnavchildmode)- high utility as it supports other actions
- [DisplayMOEW](https://wiki.cerner.com/display/MPDEVWIKI/DisplayMOEW?src=contextnavchildmode)- high utility as it supports other actions
- [GetAvailableOrderActions](https://wiki.cerner.com/display/MPDEVWIKI/GetAvailableOrderActions?src=contextnavchildmode) - moderate utility
- [GetConsolidatedRoutingOptions](https://wiki.cerner.com/display/MPDEVWIKI/GetConsolidatedRoutingOptions?src=contextnavchildmode) - unclear
- [GetDefaultRoutingDisplay](https://wiki.cerner.com/display/MPDEVWIKI/GetDefaultRoutingDisplay?src=contextnavchildmode) - unclear
- [GetScratchPadOrders](https://wiki.cerner.com/display/MPDEVWIKI/GetScratchPadOrders?src=contextnavchildmode) - low utility
- [GetXMLOrdersMOEW](https://wiki.cerner.com/display/MPDEVWIKI/GetXMLOrdersMOEW?src=contextnavchildmode) - high utility
- [InvokeCancelDCAction](https://wiki.cerner.com/display/MPDEVWIKI/InvokeCancelDCAction?src=contextnavchildmode) - inferior to MPAGES_EVENT:ORDERS
- [InvokeCompleteAction](https://wiki.cerner.com/display/MPDEVWIKI/InvokeCompleteAction?src=contextnavchildmode) - inferior to MPAGES_EVENT:ORDERS
- [InvokeRenewAction](https://wiki.cerner.com/display/MPDEVWIKI/InvokeRenewAction?src=contextnavchildmode) - inferior to MPAGES_EVENT:ORDERS
- [InvokeRenewActionWithRouting](https://wiki.cerner.com/display/MPDEVWIKI/InvokeRenewActionWithRouting?src=contextnavchildmode) - inferior to MPAGES_EVENT:ORDERS
- [InvokeResolveActionMOEW](https://wiki.cerner.com/display/MPDEVWIKI/InvokeResolveActionMOEW?src=contextnavchildmode) - unclear
- [InvokeRetailMedManagerRefillRequestReview](https://wiki.cerner.com/display/MPDEVWIKI/InvokeRetailMedManagerRefillRequestReview?src=contextnavchildmode) - unclear
- [IsScratchPadEmptyMOEW](https://wiki.cerner.com/display/MPDEVWIKI/IsScratchPadEmptyMOEW?src=contextnavchildmode) - low utility
- [RemoveOrderFromScratchPad](https://wiki.cerner.com/display/MPDEVWIKI/RemoveOrderFromScratchPad?src=contextnavchildmode) - low utility
- [SignOrders](https://wiki.cerner.com/display/MPDEVWIKI/SignOrders?src=contextnavchildmode) - high utility as it supports other actions

**Comment by geekmdtravis** [comment]
@dl2github - We'll definitely need to explore implementation options before starting. Specifically comparing the functionality of MPAGE_EVENTS:ORDERS and DiscernObjectFactor:ORDERS to determine which function to use, when, and how to modify the return types, plus how many functions might be needed (e.g. `submitOrders` and `submitPowerPlan` or just `submitOrders`).

**Comment by geekmdtravis** [comment]
This is great @dl2github! I think another useful way of breaking this down would be a comparison table. For example:

| Function | MPAGES_EVENT   | DiscernObjectFactory  |
|---|---|---|
|  Repeat (copy) Order| &check;   | -   |
|  New Order | &check;   | &check;   |

Etc.

I believe we'll eventually need to support both order pathways, but I believe it's likely to be best to have an API that abstracts away some of the more tedious facets of the native Discern functions and COM Objects and simply provides functionality. In that case, we'll have to __choose__ between MPAGES_EVENT:ORDERS and DiscernObjectFactor:POWERORDERS to perform certain tasks.

**Comment by geekmdtravis** [comment]
@geekmdtravis Great call, thank you! Updated format below:

| Order Actions | MPAGES_EVENT   | DiscernObjectFactory  | Comments |
|---|---|---|---|
|  Add Diagnoses to Order | partial |  &check;   | Can add diagnoses _with_ orders |
|  Add Powerplan Orders | &cross; |  &check;   |&mdash;|
|  Activate Order | &check;   |  &check;   |&mdash;|
|  Cancel/Discontinue Order | &check;   | &check; |&mdash;|
|  Cancel/Reorder Order | &check;   | &cross;  |&mdash|
|  Cancel/Transfer Order | &cross;   | &check;   | What does this do?|
|  Clear Order | &check;   | &check;   |&mdash;|
|  Complete Order | &cross;  |  &check;   |Is this for a proposed order?|
|  Convert to Inpatient Order | &check;   | &check;   |&mdash;|
|  Convert to Prescription Order | &check;   | &check;   |What does this do?|
|  Copy Order| &check;   | &cross;   |Very important functionality!|
|  Cosign Order| &cross;   |  &check;   |How does this differ from complete?|
|  Give With Order| &cross;   |  &check;   |What does this do?|
|  Modify Order | &check;   | &check;   |Does this actually work?|
|  New Order | &check;   | &check;   |Very important functionality.|
|  Obtain Order Comments | &cross;   | &check;   |Cool utility!|
|  Obtain Order Ingredients |&cross;   | &check;   |Cool utility!|
|  Obtain Order Reference & Information| &cross;   | &check;   |What does this do?|
|  Renew Non-Prescription Order | &check;   | &check;    |How does this differ from Rx orders?|
|  Renew Prescription Order | &check;   | &check;    |How does this differ from non-Rx orders? |
|  Reschedule Order | &cross;   | &check;   |What does this do?|
|  Resolve or Remove Order |&cross;   | &check;    |From scratchpad?|
|  Resume Order | &check;   | &check;    |&mdash;|
|  Retrieve XML Order Information | &cross;  | &check;    |Guessing this is similar to the MPAGES_EVENT return value, but need to check.|
|  Suspend Order | &check;   | &check;    |&mdash;|
|  Unlink Order | &cross;  | &check;    |What does this do?|
|  Void Order | &cross;  | &check;    |How does this differ from cancel/discontinue?|

**Comment by geekmdtravis** [comment]
@dl2github  - I think we need to create a more direct mapping of actions between the two pathways to order management. Please see the example below.

|Action|MPAGES_EVENT|DiscernObjectFactory|Comments|
|---|---|---|---|
|Copy and reorder|REPEAT|&cross;|&mdash;|
|Cancel and discontinue|CANCEL DC|InvokeCancelDCAction|&mdash;|
|Renew|RENEW|InvokeRenewAction|May not work in our implementation PowerChart, need to test|

**Comment by geekmdtravis** [comment]
@dl2github  - This looks great, thanks. Looking things over, I'm getting the impression we can still use `MPAGES_EVENT` for placing/modifying/cancelling orders without any real loss of functionality when compared to `DiscernObjectFactory`. The auxillary/utility functions in `DiscernObjectFactory` are likely all still useful though. Thoughts/

**Comment by geekmdtravis** [comment]
@dl2github  - Perfect. Then, if you can pick it up, let's have you lead the development of an `async` function that creates _Power Plans_, plus any requisite accessory functions, types, etc. Please keep me involved with API design.

**Comment by geekmdtravis** [comment]
@dl2github  - an important note if you haven't picked up on it already. To properly implement the `DiscernObjectFactory` in the `TypeScript` environment, you'll need to update the `global` type definition` declared in `src/index.ts`.

**Comment by geekmdtravis** [comment]
Merged into dev.
